### PR TITLE
Disk size check

### PR DESCRIPTION
**Pull request by vkresch**
Resolves https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/issues/504

* **Analysis**: As described in the issue TRD retries payments when the disk is full because it does not handle this edge case properly.

* **Solution**: The solution proposed in this PR is to check during the TRD run the disk size and to shut down properly (with the transition `SHUT_DOWN_ON_ERROR`) if the free disk size of the system is less then 10% ([source](https://serverfault.com/questions/10475/how-much-space-should-you-leave-free-on-a-hard-disk)). If we check everytime in the running loop the disk size we always make sure that TRD has enough disk space independent of the many write function location as described by [here](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/issues/504#issuecomment-945411517). 

I was also considering and trying out adding a new state between `PRODUCERS_READY` and `CONSUMERS_READY`:
image
but the actual while loop for writing is inside the  `PRODUCERS_READY` state.
The `PRODUCER_READY` state is an aggregated state of TRD thus in future PRs we should consider splitting up this state into multiple smaller states to come closer to a finite state machine with all the exception handlings.

* **Performed tests**: Implemented test_disk_full.py regression test. Tested on own system increasing the constant percentage threshold.

* **Check list**:

- [x] I extended the Github Actions CI test units with the corresponding tests for this new feature (if needed).
- [ ] I extended the Sphinx documentation (if needed).
- [ ] I extended the help section (if needed).
- [ ] I changed the README file (if needed).
- [ ] I created a new issue if there is further work left to be done (if needed).

**Work effort**: 7h

### CODE REVIEW

**Comment by jdsika** [comment]
I think there is some minimum free space on a system that sometimes locks out non admin processes before the disc is even full. Like e.g. 5% free space -> all programs fail but the system can still write something in order to be able to gracefull handle the nearing full disk?

**Comment by vkresch** [comment]
@jdsika yes good point I also found a discussion [here](https://serverfault.com/questions/10475/how-much-space-should-you-leave-free-on-a-hard-disk). To be on the safe side let's make it 10%. I will update the constant accordingly.

**Comment by jdsika** [comment]
@vkresch will this be ready today?

**Comment by vkresch** [comment]
I will try finishing it today but latest tomorrow.

**Comment by jdsika** [comment]
I have merged the monster PR. You can resolve conflicts here and finish!

**Comment by utdrmac** [comment]
Why isn't this failure being checked?
https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/blob/db45c466f6598b06bd89fcbd412af93b7a7c5ab1/src/util/csv_payment_file_parser.py#L31

This seems like the easiest solution. If you can't write() to a file, abort TRD completely. Checking for available disk space is a good preemptive test, but the actual writing should still send admin notification and halt TRD if writing fails.

File: src/Constants.py [hunk 1]
```diff
@@ -1,22 +1,22 @@
...
+CURRENT_TESTNET = '{}NET'.format(PROTOCOL_NAME)
 
 FIRST_GRANADA_LEVEL = 1589249
 
 # Providers api prefix
 # Public RPC
 PUBLIC_NODE_URL = {
     'MAINNET': 'https://mainnet-tezos.giganode.io',
-    CURRENT_TESTNET: 'https://granadanet.smartpy.io'
+    CURRENT_TESTNET: 'https://{}.smartpy.io'.format(CURRENT_TESTNET)
```
Comment (vkresch) [inline src/Constants.py:13]: This is the culprit. The url` https://granadanet.smartpy.io` returns 502 error. Means it has been taken down.

File: src/Constants.py [hunk 2]
```diff
@@ -1,22 +1,22 @@
 from enum import Enum
 
 EXIT_PAYMENT_TYPE = "exit"
-
-CURRENT_TESTNET = 'GRANADANET'
+PROTOCOL_NAME = 'GRANADA'
```
Comment (vkresch) [inline src/Constants.py:4]: Question: I tried updating the protocol name to `HANGZHOU` but it seems that the [current_level](https://hangzhounet.smartpy.io/chains/main/blocks/head/header/) of this protocol is at 169123. Which is significant less then in granada [current_level](https://rpc.tzkt.io/granadanet/chains/main/blocks/head/header) with 765770. 

I am asking because then it fails the `test_calculatePhase0.py` test. Because the snapshot_level (266241) is bigger then the current_level (169123) in `HANGZHOU`. See this if check [here](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/blob/master/src/rpc/rpc_reward_api.py#L482). How do we want to deal with that?

File: tests/unit/test_calculatePhase0.py [hunk 3]
```diff
@@ -15,7 +15,7 @@ def test_calculate(self):
 
         nw = DEFAULT_NETWORK_CONFIG_MAP[CURRENT_TESTNET]
 
-        api = ProviderFactory(provider='prpc').newRewardApi(nw, BAKING_ADDRESS, '')
+        api = ProviderFactory(provider='tzkt').newRewardApi(nw, BAKING_ADDRESS, '')
```
Comment (vkresch) [inline tests/unit/test_calculatePhase0.py:18]: Because https://granadanet.smartpy.io return 502 I use tzkt api.

**Comment by vkresch** [comment]
> This may not be where TRD is running, nor where the user is saving reports. This should check disk space at the path of --reports-dir as that is the failure issue. If reports cannot be saved, then TRD repeats forever.

@utdrmac yes in the issue the error occured in the payment loop when it try_to_pay and create_calculations_report. I was thinking it might be beneficial to interrupt the main loop as early as possible to avoid unnecessary calculations and api requests.

File: src/cli/client_manager.py [hunk 4]
```diff
@@ -24,8 +23,6 @@ def __init__(self, node_endpoint, signer_endpoint) -> None:
             else:
                 self.node_endpoint = 'https://' + self.node_endpoint
                 logger.info("Node endpoint URL points to an SSL endpoint. Using HTTPS protocol prefix.")
-        if len(self.node_endpoint.split(':')) < 3:
-            self.node_endpoint += f':{TEZOS_RPC_PORT}'
```
Comment (jdsika) [inline src/cli/client_manager.py:28]: Why did you remove this?

File: tests/smoke/test_launch_states.py [hunk 5]
```diff
@@ -25,9 +25,9 @@
 @pytest.fixture
 def args():
     # Test with PRPC node
-    args = Args(initial_cycle=10, reward_data_provider='tzkt', api_base_url='https://api.carthage.tzkt.io/v1/')
+    args = Args(initial_cycle=10, reward_data_provider='tzkt', api_base_url='https://api.hangzhounet.tzkt.io/v1/')
```
Comment (jdsika) [inline tests/smoke/test_launch_states.py:28]: This needs to be the URL from constants.py ? it was hangzhou2net and I changed it

File: tests/smoke/test_launch_states.py [hunk 6]
```diff
@@ -25,9 +25,9 @@
...
     args.network = 'EDO2NET'
```
Comment (jdsika) [inline tests/smoke/test_launch_states.py:29]: Needs change. I think we should run all tests here on mainnet and not testnet, right?

File: tests/smoke/test_tzkt_api.py [hunk 7]
```diff
@@ -58,5 +58,5 @@ def test_base_url(ConfigParser, args):
         )
     )
     args.initial_cycle = 100
-    args.api_base_url = 'https://api.carthage.tzkt.io/v1/'
+    args.api_base_url = 'https://api.tzkt.io/v1/'
```
Comment (jdsika) [inline tests/smoke/test_tzkt_api.py:61]: This is the mainnet URL and above is the testnet URL from giganode. this doesn't seem to fit

**Comment by jdsika** [comment]
Are you checking the disk space of --reports-dir as well?

File: tests/unit/test_calculatePhase0.py (hunk 3 above)
Comment (jdsika) [reply tests/unit/test_calculatePhase0.py]: I updated the the testnet URL

File: tests/regression/test_disk_full.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+)
+
+
+@pytest.fixture
+def args():
+    # Test with PRPC node
+    args = Args(
+        initial_cycle=10,
+        reward_data_provider="tzkt",
+        api_base_url="https://api.tzkt.io/v1/",
```
Comment (jdsika) [inline tests/regression/test_disk_full.py:37]: I don't want any URLs in the tests. They should all be taken from constants.py! We are also again mixing testnet and mainnet. I would vote to do the tests only on mainner and not on testnet if there are especially tests that invole a specific block level etc

File: src/cli/client_manager.py (hunk 4 above)
Comment (vkresch) [reply src/cli/client_manager.py:28]: The tezos rpc port url does not work anymore. Thus I removed it.

**Comment by nicolasochem** [comment]
I have the same comment than @jdsika , we need to make sure that the partition used for reports_dir is the one checked for usage. [This](https://stackoverflow.com/a/12327880/207209) looks like a good method to do it in python.

File: src/Constants.py (hunk 2 above)
Comment (jdsika) [reply src/Constants.py]: We will make tests with mainnet in order to avoid that trouble

**Comment by jdsika** [comment]
@denver-s I need your help here quickly:

```
DEBUG    verbose:rpc_reward_api.py:245 [do_rpc_request] Requesting URL https://mainnet-tezos.giganode.io/chains/main/blocks/252928/context/delegates/tz1RpNqAzQNaZuxeLSbfEzVEzUWssddsL6Kw
ERROR    main.rpc_reward_api:rpc_reward_api.py:427 [get_d_d_b] RPC API Error: [do_rpc_request] Requesting URL 'https://mainnet-tezos.giganode.io/chains/main/blocks/252928/context/delegates/tz1RpNqAzQNaZuxeLSbfEzVEzUWssddsL6Kw' failed (500)
=========================== short test summary info ============================
FAILED tests/unit/test_calculatePhase0.py::TestCalculatePhase0::test_calculate
```

You added this baker address but it does not exist. I changed the network to mainnet as the testnetwork does always reboot and start at level 0. I am not sure what the test intends to do and would like to ask you to add a proper baker address that will not fail with future protocol upgrades. Thank you!

**Comment by denver-s** [comment]
@jdsika Previously, it was using [another testnet addres](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/blob/53693ab0942357fb9a88164d16c1dbd7197628c6/tests/unit/test_calculatePhase0.py) so I just changed to a Granadanet address. I am sure that it's better to use a mainnet address to avoid such errors.

**Review by jdsika — APPROVED** [review]
